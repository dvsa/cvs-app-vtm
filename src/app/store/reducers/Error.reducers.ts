import { initialAppState, IAppState } from './../state/app.state';
import { ErrorActions, EErrorActions } from '../actions/Error.actions';

export function ErrorReducer(state = initialAppState.error, action: ErrorActions):
  IAppState['error'] {
  switch (action.type) {
    case EErrorActions.SetErrorMessage: {
      return action.payload;
    }

    case EErrorActions.ClearErrorMessage: {
      return null
    }

    default: return state;
  }
}
