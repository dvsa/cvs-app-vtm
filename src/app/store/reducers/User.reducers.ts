import * as fromUser from '../actions/User.action';
import { UserDetails } from "@app/models/user-details";

export interface UserState {
  details: UserDetails
}
// UserDetails
export const initialUserState: UserState = {
  // oid & unique_name
  details: null
}

export function UserReducer(
  state: UserState = initialUserState,
  action: fromUser.UserAction
):  UserState {

  // @ts-ignore does exist, throws error as action is imported as *
  const {type, payload} = action;
  switch(type) {
    // will dispatch AppIsLoading
    case fromUser.SET_USER: {
      return {
        ...state
      };
    }
    case fromUser.SET_USER_SUCCESS: {
      return {
        ...state,
        details: {
          ...payload
        }
      };
    }
    // will dispatch loading AppIsNotLoading
    case fromUser.SET_USER_FAIL: {
      return {
        ...state,
      };
    }
    default:
      return state;
  } 
}

export const getUserDetails = ({details}: UserState) => details;
