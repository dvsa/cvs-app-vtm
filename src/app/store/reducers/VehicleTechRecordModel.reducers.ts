import { initialVehicleTechRecordModelState, IVehicleTechRecordModelState } from '@app/store/state/VehicleTechRecordModel.state';
import { EVehicleTechRecordModelActions, VehicleTechRecordModelActions } from '@app/store/actions/VehicleTechRecordModel.actions';
import { IAppState } from '@app/store/state/app.state';

export function VehicleTechRecordModelReducers(
  state: IVehicleTechRecordModelState = initialVehicleTechRecordModelState,
  action: VehicleTechRecordModelActions): IVehicleTechRecordModelState {
  switch (action.type) {
    case EVehicleTechRecordModelActions.GetVehicleTechRecordModelSuccess: {
      return {
        ...state,
        selectedVehicleTechRecordModel: action.payload
      };
    }

    case EVehicleTechRecordModelActions.GetVehicleTechRecordModelHavingStatusAllSuccess: {
      return {
        ...state,
        vehicleTechRecordModel: action.payload
      };
    }

    default:
      return state;
  }
}
