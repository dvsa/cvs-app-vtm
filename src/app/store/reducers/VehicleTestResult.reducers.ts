import { initialVehicleTestResultModelState } from '@app/store/state/VehicleTestResult.state';
import {
  EVehicleTestResultActions,
  VehicleTestResultActions
} from '../actions/VehicleTestResult.actions';

export function VehicleTestResultReducers(
  state = initialVehicleTestResultModelState,
  action: VehicleTestResultActions
) {
  switch (action.type) {
    case EVehicleTestResultActions.GetVehicleTestResultModelSuccess: {
      return {
        ...state,
        vehicleTestResultModel: action.payload
      };
    }

    case EVehicleTestResultActions.GetVehicleTestResultModelFailure: {
      return {
        ...state,
        vehicleTestResultModel: null
      };
    }

    case EVehicleTestResultActions.SetTestViewState: {
      return {
        ...state,
        editState: action.editState
      };
    }

    case EVehicleTestResultActions.UpdateTestResultSuccess: {
      return {
        ...state,
        vehicleTestResultModel: action.testResultTestTypeNumber.testResultsUpdated,
        selectedTestResultModel: action.testResultTestTypeNumber.testResultUpdated
      };
    }

    case EVehicleTestResultActions.SetSelectedTestResultModelSuccess: {
      return {
        ...state,
        selectedTestResultModel: action.payload[0]
      };
    }

    case EVehicleTestResultActions.UpdateSelectedTestResultModelSuccess: {
      return {
        ...state,
        selectedTestResultModel: action.payload
      };
    }

    case EVehicleTestResultActions.CreateTestResultSuccess: {
      return {
        ...state,
        selectedTestResultModel: action.vTestResultUpdated
      };
    }

    default:
      return state;
  }
}
