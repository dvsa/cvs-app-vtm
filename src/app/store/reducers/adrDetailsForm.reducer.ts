import {Action, combineReducers} from '@ngrx/store';
import {
  AddArrayControlAction,
  addGroupControl,
  createFormStateReducerWithUpdate,
  RemoveArrayControlAction,
  setValue,
  updateGroup,
  validate
} from 'ngrx-forms';
import {greaterThan, lessThanOrEqualTo, maxLength, required} from 'ngrx-forms/validation';
import {adrDetailsFormModel} from '@app/models/adrDetailsForm.model';
import {IAppState, INITIAL_STATE} from '../state/adrDetailsForm.state';
import {
  CreateGuidanceNoteElementAction,
  CreatePermittedDangerousGoodElementAction,
  CreateProductListUnNoElementAction, CreateTc3TypeElementAction,
  RemoveGuidanceNoteElementAction,
  RemovePermittedDangerousGoodElementAction,
  RemoveProductListUnNoElementAction,
  SetSubmittedValueAction
} from '../actions/adrDetailsForm.actions';
import {approvalDate} from '@app/models/approvalDate';

const formGroupReducerWithUpdate = createFormStateReducerWithUpdate<adrDetailsFormModel>(updateGroup<adrDetailsFormModel>({
  name: validate(required, maxLength(150)),
  street: validate(required, maxLength(150)),
  town: validate(required, maxLength(100)),
  city: validate(required, maxLength(100)),
  postcode: validate(required, maxLength(25)),
  type: validate(required),
  approvalDate: updateGroup<approvalDate>({
    day: validate(required, greaterThan(0), lessThanOrEqualTo(31)),
    month: validate(required, greaterThan(0), lessThanOrEqualTo(12)),
    year: validate(required, greaterThan(1970)),
  }),
  permittedDangerousGoods: validate(required),
  additionalNotes: validate(required),
  tankManufacturer: validate(maxLength(70)),
  tankManufacturerSerialNo: validate(maxLength(50)),
  tankTypeAppNo: validate(maxLength(65)),
  tankCode: validate(maxLength(30)),
  statement: validate(maxLength(1500)),
  productList: validate(maxLength(1500)),
  specialProvisions: validate(maxLength(1024)),
  tc2IntermediateApprovalNo: validate(maxLength(75)),
  tc3PeriodicNumber: validate(maxLength(75)),
  memosApply: validate(required),
  batteryListNumber: validate(maxLength(8))

}));


const reducers = combineReducers<IAppState['adrDetails'], any>({
  formState(
    s = INITIAL_STATE,
    a: CreatePermittedDangerousGoodElementAction | RemovePermittedDangerousGoodElementAction |
       CreateGuidanceNoteElementAction | RemoveGuidanceNoteElementAction |
       CreateProductListUnNoElementAction | RemoveProductListUnNoElementAction |
       CreateTc3TypeElementAction
  ) {
    s = formGroupReducerWithUpdate(s, a);

    switch (a.type) {
      case CreatePermittedDangerousGoodElementAction.TYPE:
        return updateGroup<adrDetailsFormModel>({
          permittedDangerousGoods: group => {
            return addGroupControl(group, a.name, false);
          },
        })(s);

      case RemovePermittedDangerousGoodElementAction.TYPE:
        return updateGroup<adrDetailsFormModel>({
          permittedDangerousGoods: group => {
            const newValue = { ...group.value };
            delete newValue[a.name];
            return setValue(group, newValue);
          },
        })(s);

      case CreateGuidanceNoteElementAction.TYPE:
        return updateGroup<adrDetailsFormModel>({
          additionalNotes: group => {
            return addGroupControl(group, a.name, false);
          },
        })(s);

      case RemoveGuidanceNoteElementAction.TYPE:
        return updateGroup<adrDetailsFormModel>({
          additionalNotes: group => {
            const newValue = { ...group.value };
            delete newValue[a.name];
            return setValue(group, newValue);
          },
        })(s);


      case CreateProductListUnNoElementAction.TYPE:
        return updateGroup<adrDetailsFormModel>({
          productListUnNo: group => {
            const newValue = { ...group.value };
            delete newValue[a.name];
            return setValue(group, newValue);
          },
        })(s);

      case RemoveProductListUnNoElementAction.TYPE:
        return updateGroup<adrDetailsFormModel>({
          productListUnNo: group => {
            const newValue = { ...group.value };
            delete newValue[a.name];
            return setValue(group, newValue);
          },
        })(s);

      default:
        return s;
    }
  },
  permittedDangerousGoodsOptions(
    s: string[] = [],
    a: CreatePermittedDangerousGoodElementAction | RemovePermittedDangerousGoodElementAction
  ) {
    // console.log(`inside reducer permittedDangerousGoodsOptions s => ${JSON.stringify(s)}, a => ${JSON.stringify(a)}`);
    switch (a.type) {
      case CreatePermittedDangerousGoodElementAction.TYPE:
        return [...s, a.name];

      case RemovePermittedDangerousGoodElementAction.TYPE:
        return s.filter(i => i !== a.name);

      default:
        return s;
    }
  },
  additionalNotesOptions(
    s: string[] = ['Document Option 1', 'Document Option 2'],
    a: CreateGuidanceNoteElementAction | RemoveGuidanceNoteElementAction
  ) {
    // console.log(`inside reducer additionalNotesOptions s => ${JSON.stringify(s)}, a => ${JSON.stringify(a)}`);
    switch (a.type) {
      case CreateGuidanceNoteElementAction.TYPE:
        return [...s, a.name];

      case RemoveGuidanceNoteElementAction.TYPE:
        return s.filter(i => i !== a.name);

      default:
        return s;
    }
  },
  productListUnNo: function (
    s = {maxIndex: 0, options: []},
    a: AddArrayControlAction<number> | RemoveArrayControlAction,
  ) {
    switch (a.type) {
      case AddArrayControlAction.TYPE: {
        console.log(`inside reducer productListUnNo AddArrayControlAction.TYPE s => ${JSON.stringify(s)}, a => ${JSON.stringify(a)}`);
        const maxIndex = s.maxIndex + 1;
        const options = [...s.options];
        options.splice(a.index!, 0, maxIndex);
        return {
          maxIndex,
          options,
        };
      }

      case RemoveArrayControlAction.TYPE: {
        const options = [...s.options];
        options.splice(a.index!, 1);
        return {
          ...s,
          options,
        };
      }

      default:
        return s;
    }
  },
  tc3Type: function (
    s = {maxIndex: 0, options: []},
    a: CreateTc3TypeElementAction,
  ) {
    switch (a.type) {
      case CreateTc3TypeElementAction.TYPE: {
        console.log(`inside reducer tc3Type AddArrayControlAction.TYPE s => ${JSON.stringify(s)}, a => ${JSON.stringify(a)}`);
        const maxIndex = s.maxIndex + 1;
        const options = [...s.options];
        // tslint:disable-next-line:no-unnecessary-type-assertion no-non-null-assertion
        options.splice(a.index!, 0, maxIndex);
        return {
          maxIndex,
          options,
        };
      }

      // case CreateTc3TypeElementAction.TYPE: {
      //   const options = [...s.options];
      //   // tslint:disable-next-line:no-unnecessary-type-assertion no-non-null-assertion
      //   options.splice(a.index!, 1);
      //   return {
      //     ...s,
      //     options,
      //   };
      // }

      default:
        return s;
    }
  },
  submittedValue(s: adrDetailsFormModel | undefined, a: SetSubmittedValueAction) {
    switch (a.type) {
      case SetSubmittedValueAction.TYPE:
        return a.submittedValue;

      default:
        return s;
    }
  },
});


export function adrDetailsReducer(state: IAppState['adrDetails'], action: Action) {
  return reducers(state, action);
}
