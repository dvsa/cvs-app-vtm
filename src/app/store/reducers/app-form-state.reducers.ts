import { AppFormStateActions, EAppFormStateActions  } from '../actions/app-form-state.actions';

export interface AppFormState {
  pristine: boolean;
}

export const initialAppFormState: AppFormState = {
  pristine: true,
};

export function AppFormStateReducer(
  state: AppFormState = initialAppFormState,
  action: AppFormStateActions
): AppFormState {
  switch (action.type) {
    case EAppFormStateActions.SetAppFormDirty: {
      return {
        ...state,
        pristine: false
      };
    }

    case EAppFormStateActions.SetAppFormPristine: {
      return {
        ...state,
        pristine: true
      };
    }

    default:
      return state;
  }
}
