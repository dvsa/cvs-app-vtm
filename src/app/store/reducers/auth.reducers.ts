import * as authActions from '../actions/auth.actions';

export interface State {
  userName?: string;
  friendlyName?: string;
}

export const initialState: State = {
  userName: null,
  friendlyName: null
};

export function reducer(state = initialState, action: authActions.AuthActions): State {
  switch (action.type) {
    case authActions.AuthActionTypes.SetAuths:
      return handleSetAuths(state, action);

    default:
      return state;
  }
}

function handleSetAuths(state: State, action: authActions.SetAuths): State {
  return {
    ...state,
    userName: action.payload.userName,
    friendlyName: action.payload.friendlyName
  };
}

export const getUserName = (state: State) => state.userName;
export const getFriendlyName = (state: State) => state.friendlyName;
