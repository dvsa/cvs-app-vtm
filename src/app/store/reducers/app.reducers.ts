import { ActionReducerMap } from '@ngrx/store';
import { routerReducer } from '@ngrx/router-store';

import { IAppState } from '@app/store/state/app.state';
import { VehicleTechRecordModelReducers } from './VehicleTechRecordModel.reducers';
import { VehicleTestResultModelReducers } from './VehicleTestResultModel.reducers';
import { LoaderReducer } from './Loader.reducers';
import { ErrorReducer } from './error.reducers';
import { ReferenceDataReducers } from '@app/store/reducers/ReferenceData.reducers';

export const appReducers: ActionReducerMap<IAppState> = {
  router: routerReducer,
  loader: LoaderReducer,
  vehicleTechRecordModel: VehicleTechRecordModelReducers,
  vehicleTestResultModel: VehicleTestResultModelReducers,
  referenceData: ReferenceDataReducers,
  error: ErrorReducer
};
