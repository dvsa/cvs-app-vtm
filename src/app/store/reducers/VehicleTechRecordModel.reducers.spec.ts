import {VehicleTechRecordModelReducers} from './VehicleTechRecordModel.reducers';
import {initialVehicleTechRecordModelState} from '../state/VehicleTechRecordModel.state';
import {
  GetVehicleTechRecordModel,
  GetVehicleTechRecordModelSuccess,
  GetVehicleTechRecordModelHavingStatusAll,
  GetVehicleTechRecordModelHavingStatusAllSuccess
} from '../actions/VehicleTechRecordModel.actions';

describe('VehicleTechRecordModel Reducer', () => {
  describe('undefined action', () => {
    it('should return the default state', () => {
      const action = {type: 'NOOP'} as any;
      const result = VehicleTechRecordModelReducers(undefined, action);

      expect(result).toBe(initialVehicleTechRecordModelState);
    });
  });

  describe('[VehicleTechRecordModel] Get VehicleTechRecordModel', () => {
    it('should toggle loading state', () => {
      const action = new GetVehicleTechRecordModel({techRecord: []});
      const result = VehicleTechRecordModelReducers(initialVehicleTechRecordModelState, action);
      expect(result).toEqual({
        ...initialVehicleTechRecordModelState
      });
    });
  });

  describe('[VehicleTechRecordModel] Get VehicleTechRecordModel Success', () => {
    it('should update the state with the payload', () => {
      const vehicleTechRecord = {
        vrms: [
          {
            vrm: '53654645',
            isPrimary: true
          }],
        vin: '1234566',
        techRecord: []
      };
      const action = new GetVehicleTechRecordModelSuccess(vehicleTechRecord);
      const result = VehicleTechRecordModelReducers(initialVehicleTechRecordModelState, action);
      expect(result).toEqual({
        ...initialVehicleTechRecordModelState,
        selectedVehicleTechRecordModel: vehicleTechRecord
      });
    });
  });

  describe('[VehicleTechRecordModel] Get VehicleTechRecordModelHavingStatusAll', () => {
    it('should toggle loading state', () => {
      const action = new GetVehicleTechRecordModelHavingStatusAll({techRecord: []});
      const result = VehicleTechRecordModelReducers(initialVehicleTechRecordModelState, action);
      expect(result).toEqual({
        ...initialVehicleTechRecordModelState
      });
    });
  });

  describe('[VehicleTechRecordModel] Get VehicleTechRecordModelHavingStatusAll Success', () => {
    it('should update the state with the payload', () => {
      const vehicleTechRecord = {
        vrms: [
          {
            vrm: '53654645',
            isPrimary: true
          }],
        vin: '1234566',
        techRecord: []
      };
      const action = new GetVehicleTechRecordModelHavingStatusAllSuccess(vehicleTechRecord);
      const result = VehicleTechRecordModelReducers(initialVehicleTechRecordModelState, action);
      expect(result).toEqual({
        ...initialVehicleTechRecordModelState,
        vehicleTechRecordModel: vehicleTechRecord
      });
    });
  });
});
