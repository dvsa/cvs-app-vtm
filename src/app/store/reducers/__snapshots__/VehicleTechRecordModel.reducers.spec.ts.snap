// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VehicleTechRecordModelReducers GetVehicleTechRecordHavingStatusAllSuccess should set vehicle record(s) to the store 1`] = `
Object {
  "vehicleTechRecordModel": Array [
    Object {
      "systemNumber": "111363",
      "techRecord": Array [
        Object {
          "statusCode": "current",
        },
      ],
      "vin": "ABCD12345",
    },
  ],
}
`;

exports[`VehicleTechRecordModelReducers SetSelectedVehicleTechRecordSuccess should set selected vehicle record to the store 1`] = `
Object {
  "selectedVehicleTechRecord": Object {
    "systemNumber": "111363",
    "techRecord": Array [
      Object {
        "statusCode": "current",
      },
    ],
    "vin": "ABCD12345",
  },
}
`;

exports[`VehicleTechRecordModelReducers SetViewState should set current viewstate to the store 1`] = `
Object {
  "viewState": 1,
}
`;

exports[`VehicleTechRecordModelReducers UpdateVehicleTechRecordSuccess should update the vehicle record in the store 1`] = `
Object {
  "vehicleTechRecordModel": Array [
    Object {
      "systemNumber": "111363",
      "techRecord": Array [
        Object {
          "createdAt": "2020-06-26T10:26:54.903Z",
        },
      ],
      "vin": "ABCD12345",
    },
  ],
}
`;

exports[`VehicleTechRecordModelReducers when initialised should set the default state 1`] = `
Object {
  "selectedVehicleTechRecord": null,
  "vehicleTechRecordModel": null,
  "viewState": 0,
}
`;
